test/VaultRegistry.test.ts

```diff
@@ -1,445 +1,943 @@
+                        "ProtocolGovernance",
+                        singleton
+                    );
+                    return this.subject;
+                }
+            );
+        });
+
+        beforeEach(async () => {
+            await this.deploymentFixture();
+            this.startTimestamp = now();
+            await sleepTo(this.startTimestamp);
+        });
+
+        describe("#constructor", () => {
+            it("creates VaultRegistry", async () => {
+                expect(ethers.constants.AddressZero).to.not.eq(
+                    this.subject.address
+                );
+            });
+            it("initializes ProtocolGovernance address", async () => {
+                expect(await this.subject.protocolGovernance()).to.be.equal(
+                    this.protocolGovernance.address
+                );
+            });
+            it("initializes ERC721 token name", async () => {
+                expect(await this.subject.name()).to.be.equal("Test");
+            });
+            it("initializes ERC721 token symbol", async () => {
+                expect(await this.subject.symbol()).to.be.equal("TST");
+            });
+        });
+
+        describe("#vaults", () => {
+            it("returns all registered vaults", async () => {
+                expect(await this.subject.vaults()).to.deep.equal([
+                    this.erc165Mock.address,
+                ]);
+                await this.anotherERC165Mock.allowInterfaceId(
+                    VAULT_INTERFACE_ID
+                );
+                this.subject
+                    .connect(this.allowedRegisterVaultSigner)
+                    .registerVault(
+                        this.anotherERC165Mock.address,
+                        this.ownerSigner.address
+                    );
+                expect(await this.subject.vaults()).to.deep.equal([
+                    this.erc165Mock.address,
+                    this.anotherERC165Mock.address,
+                ]);
+            });
+
+            describe("access control:", () => {
+                it("allowed: any address", async () => {
+                    await withSigner(randomAddress(), async (s) => {
+                        await expect(this.subject.connect(s).vaults()).to.not.be
+                            .reverted;
+                    });
+                });
+            });
+        });
+
+        describe("#vaultForNft", () => {
+            it("resolves Vault address by VaultRegistry NFT", async () => {
+                expect(await this.subject.vaultForNft(this.nft)).to.equal(
+                    this.erc165Mock.address
+                );
+                expect(await this.subject.vaultForNft(this.randomNft)).to.equal(
+                    ethers.constants.AddressZero
+                );
+                await this.anotherERC165Mock.allowInterfaceId(
+                    VAULT_INTERFACE_ID
+                );
+                this.randomNft = Number(
```

Comment:
Looks like this should be a local variable

Reply:
Reassigning to this NFT would defeat the purpose of it being random and break the semantics of other test cases

Reply:
To avoid these kind of situations I'd rather make a helper randomNft = () => Math.random(...), use it everywhere in test cases and remove randomNft from context